// yo3/_ykit.qreact.js
exports.config = function (ykitPlugin) {
    return {
        exports: [
            './component_dev/index.js'
        ],
        modifyWebpackConfig: function (config) {
            var plugin = new this.webpack.DefinePlugin({
                'process.env': {
                    NODE_ENV: JSON.stringify('dev')
                }
            });
            var outputConfig = {
                path: 'dist',
                library: 'Yo',
                libraryTarget: 'umd'
            };
            config.context = './';
            config.plugins.push(plugin);

            config.output.dev = Object.assign(config.output.dev, outputConfig, {
                filename: 'yo3.qreact[ext]'
            });
            config.output.prd = Object.assign(config.output.prd, outputConfig, {
                filename: 'yo3.qreact.min[ext]'
            });
            config.externals = {
                react: 'React',
                'react-dom': 'ReactDOM'
            };
            config.resolve = {
                alias: {
                    react: 'qreact',
                    'react-dom': 'qreact',
                    'prop-types': 'qreact/lib/ReactPropTypes',
                    'create-react-class': 'qreact/lib/createClass',
                    'react-tap-event-plugin': 'qreact/lib/injectTapEventPlugin.js'
                }
            };
            config.module = {
                loaders: [
                    {
                        test: /\.scss$/,
                        loader: ykitPlugin.ExtractTextPlugin.extract(['css', 'sass'])
                    },
                    {
                        test: /\.js$/,
                        exclude: /node_modules/,
                        loader: 'babel',
                        query: {
                            presets: ['es2015', 'react', 'stage-0']
                        }
                    }
                ]

            };

            config.devtool = 'source-map';

            // ykit 的一个配置，用来实现 .scss -> .css
            // config.entryExtNames.css.push('.scss')

            return config;
        }
    };
};
